### PR TITLE
minor cleanup and refactoring for consistency and simplicity

### PR DESCRIPTION
**Pull request by joelanford**
Somewhat scatter-brained refactor/fixup PR for random things I found. I'll clean this up more if desired.

Signed-off-by: Joe Lanford <joe.lanford@gmail.com>

### CODE REVIEW

**Comment by timflannagan** [comment]
Looks reasonable to me at a glance. This will fail the linting check due to the package import grouping, and I added some unit tests for when the TypeApplied condition is nil, so you'd have to update those as well.

File: internal/util/util.go
Comment (kuiwang02) [inline internal/util/util.go:99]: @joelanford Hi, when `equality.Semantic.DeepEqual(po.Status, platformv1alpha1.PlatformOperatorStatus{})` return True, does it mean the PO status part is empty? 
if it is correct understanding, I can not understand why `return nil` here because I think`return nil` means PO is in healthy status in method `InspectPlatformOperators `, and not report error to aggregated CO. But we do not the PO final status yet (maybe finally the PO is not installed successfully). So, here maybe `return buildPOFailureMessage(po.GetName(), "PO has no status field")`

(I just worry about that before PO installation is done (no final status yet), we report wrong information to aggregated CO. Thanks to check my concern)

Comment (joelanford) [reply internal/util/util.go:99]: > does it mean the PO status part is empty?

Yes

> I can not understand why return nil here 

@timflannagan and I were chatting about this yesterday. In general, a PO with no status is a PO that has yet to be reconciled by the PO controller and that typically happens only as the PO is _just_ created. When the PO controller reconciles this PO, it'll attach a status and then this controller will reconcile again.

One thought I had to guard against the possibility of the PO controller never actually reconciling the PO: We could check `status is empty` and `creationTimestamp is within the last 1 minute`.

Which would result in:
1. Empty status within a minute of being created: no problem, return `nil`
2. Empty status after a minute of being create: something's wrong, return an error.

WDYT?

File: controllers/aggregated_clusteroperator_controller.go
Comment (tylerslaton) [inline controllers/aggregated_clusteroperator_controller.go:84]: We have some condition constants already defined but it doesn't seem like we are using them here for some reason. Maybe we just need to add onto that list?

https://github.com/openshift/platform-operators/blob/03149b26e54fa0737b37a9e183bb75fbf73ea94a/api/v1alpha1/types.go#L22-L25

File: controllers/aggregated_clusteroperator_controller.go
Comment (tylerslaton) [inline controllers/aggregated_clusteroperator_controller.go:96]: In general should we be using the `apimachinery` api's for this instead of the `OpenShift` ones? Mainly asking to see how you decided to change this.

**Review by tylerslaton — APPROVED** [review]
Nice work!

/lgtm

File: internal/util/util.go
Comment (timflannagan) [reply internal/util/util.go:99]: @joelanford Should we add a comment that basically explains what you had written down here? I should've done that in the PR that refactored these checks.

In general, I think that option you outlined seems reasonable at a glance but introduces potential complexity.

File: internal/util/util.go
Comment (timflannagan) [inline internal/util/util.go:91]: :+1: Whoops. This was a bit obvious in hindsight :sweat_smile:.

**Comment by timflannagan** [comment]
lgtm

File: internal/util/util.go
Comment (timflannagan) [reply internal/util/util.go:99]: Thinking about this more, in the current implementation I think we'll see flapping in the aggregate ClusterOperator resource we're managing. I'm working on the POM component's implementation, which is responsible for bubbling up the underlying BundleDeployment status. In the case where the BD resource is waiting to be unpacked, we'll effectively see there's no TypeApplied status condition type present yet, and propagate the ApplyPending status condition reason in the PO resource. The aggregate CO controller would see that, and mark the aggregate CO resource with the following:

```bash
$ k get co -w
...
platform-operators-aggregated                        False       True          False      0s      encountered the failing cincinnati-operatorrwxr4 platform operator with reason "ApplyPending"
...
platform-operators-aggregated                        False       True          False      0s      encountered the failing cincinnati-operatorrwxr4 platform operator with reason "ApplyPending"
platform-operators-aggregated                        True        True          False      0s      
```

Which feels like the wrong behavior, UX, etc.

File: controllers/aggregated_clusteroperator_controller.go
Comment (joelanford) [reply controllers/aggregated_clusteroperator_controller.go:84]: Those are reasons for the PO `Applied` condition type. These are reasons for the ClusterOperator condition types.

File: controllers/aggregated_clusteroperator_controller.go
Comment (joelanford) [reply controllers/aggregated_clusteroperator_controller.go:96]: They're both definitions of basically the exact same thing. I changed this mainly because apimachinery has some nice helpers for interacting with conditions that helps us avoid maintaining complexity around the "rules" of conditions, like knowing when to bump last transition time.

File: internal/util/util.go
Comment (kuiwang02) [reply internal/util/util.go:99]: @joelanford Thanks to check my comment.

Your idea is ok, but potential complexity (same to timflannagan).
maybe we just change `return nil` to `return buildPOFailureMessage(po.GetName(), platformtypes.ReasonApplyPending)` because we possible treat `no reconciled yet` as `ApplyPending`.

anyway I am also OK if you want to take your idea on 1minute.

Comment (joelanford) [reply internal/util/util.go]: My main concern is that it seems misleading and alarmist to say a PO is failing when in fact it has just been created and is waiting for the PO controller to do something with it.

In the most recent commit, I've updated to include the "time since creationTimestamp" check. But if there's another approach that lets us avoid that complexity AND avoid saying the PO is failing, I'm totally willing to get rid of that.

File: internal/util/util.go
Comment (joelanford) [inline internal/util/util.go:95]: Just want to highlight that my rebase included this addition based on the earlier discussion here: https://github.com/openshift/platform-operators/pull/40#discussion_r964424982

**Comment by timflannagan** [comment]
The e2e results are a no-op until #58 merges.

**Comment by timflannagan** [comment]
Now that the payload PR has merged, does it make sense to track this as a bug somewhere? I'm not sure we have a decent Jira home right now, so maybe an upstream issue could be sufficient here.

**Comment by timflannagan** [comment]
/retitle OLM-2776 minor cleanup and refactoring for consistency and simplicity

**Comment by timflannagan** [comment]
/retitle OCPBUGS-1477 minor cleanup and refactoring for consistency and simplicity

**Comment by joelanford** [comment]
/hold

I need to think through this CVO doc line item before this merges:

> An operator should not report the Available status condition the first time until they are completely rolled out (or within some reasonable percentage if the component must be installed to all nodes)

If we consider just-created, but not-yet-reconciled POs to be Available, we'll signal to CVO that the POs have been "completely rolled out" prior to the rollout even starting. It seems like we may need some logic that:
1. During a cluster install scenario, ensures all POs have BDs that have progressed to `InstallationSucceeded` before setting A=T
2. After initial cluster install, any newly created POs should trigger `Progressing=true` until that newly created PO's BD has attempted an install.

**Comment by joelanford** [comment]
I reverted the changes to the way the Available condition is calculated. For now, that means we'll keep the existing logic which means that a PO that is missing the `Installed` condition will be considered Available=False.

**Comment by joelanford** [comment]
I also disassociated this PR from OCPBUGS-1477 since it no longer addresses that bug. And I actually don't think that's a bug in the first place. So I think I'd suggest that we capture any takeaways and close it as "not a bug"

File: controllers/aggregated_clusteroperator_controller.go
Comment (timflannagan) [inline controllers/aggregated_clusteroperator_controller.go:110]: I've seen a bunch of other core ClusterOperator resources set the "AsExpected" reason during success states.

File: controllers/aggregated_clusteroperator_controller.go
Comment (timflannagan) [inline controllers/aggregated_clusteroperator_controller.go:86]: We could also move this logic to a method/function that adds these related objects to the builder instance.

**Comment by timflannagan** [comment]
The refactoring generally lgtm. It would be nice to add constants for the common set of reasons that we expect to use here.

**Comment by joelanford** [comment]
Latest commit covers your latest round of feedback @timflannagan.

**Comment by joelanford** [comment]
> Another potential TODO is updating the aggregate CO implementation to re-create that resource when it no longer exists.

Based on what Trevor said yesterday, I kinda think we leave it up to CVO to recreate the CO.

File: controllers/aggregated_clusteroperator_controller.go
Comment (tylerslaton) [inline controllers/aggregated_clusteroperator_controller.go:67]: nit:
```suggestion
			log.Error(err, "error updating ClusterOperator status")
```

All other instances apply if we want to change this.

File: controllers/aggregated_clusteroperator_controller.go
Comment (tylerslaton) [inline controllers/aggregated_clusteroperator_controller.go:97]: Maybe we could define these two conditions as reasons and then bubble them up via `util.InspectPlatformOperators()` return? That way we can set that reason inside this `WithAvailable()` call and 
have a less generic "PlatformOperatorError" reason.

**Comment by timflannagan** [comment]
Looks like some of the e2e tests need to be updated too.

/lgtm cancel

**Comment by timflannagan** [comment]
```
{Timed out after 60.001s.
Expected
    <string>: All platform operators are in a successful state
to contain substring
    <string>: All platform operators in a successful state failed /go/src/github.com/openshift/platform-operators/test/e2e/aggregated_clusteroperator_test.go:111
```

Looks like we're missing a couple more test cases :upside_down_face:.

**Comment by joelanford** [comment]
🤮 this is what I get for trying to fix these without actually looking too closely at the failures.

**Comment by joelanford** [comment]
Meta-question: Is there a process I'm not aware of to merge non-bug/non-feature PRs? The label requirements seem to imply one or the other.

**Comment by timflannagan** [comment]
We're doing the TE slides so manually adding the px-approved label after a slack conversation in the #tmp-platform-operators channel.

/label px-approved

Holding, as I think we need QE approval give we're modifying the expected status condition reason/messages/etc. and it's unclear whether that will break any automated testing.

/hold

**Comment by timflannagan** [comment]
This should also be a no-op for docs. Manually adding that label.

/label docs-approved

**Comment by timflannagan** [comment]
Superseding this in #75 given the merge conflicts.

/close

**Pull request closed without merging**
